Fix slider path placement snapping non-head nodes to distanced snapping grid

peppy opened the pull request:
As discussed at https://github.com/ppy/osu/discussions/23531.

peppy committed:
Split out grid snapping modes into "relative" and "global" types

peppy committed:
Fix slider path placement snapping non-head nodes to distanced snapping grid

As discussed at https://github.com/ppy/osu/discussions/23531.

bdach commented on osu.Game.Rulesets.Catch/Edit/CatchHitObjectComposer.cs:
```diff
@@ -132,7 +132,7 @@ public override SnapResult FindSnappedPositionAndTime(Vector2 screenSpacePositio
             result.ScreenSpacePosition.X = screenSpacePosition.X;
 
-            if (snapType.HasFlagFast(SnapType.Grids))
+            if (snapType.HasFlagFast(SnapType.GlobalGrids))
```
This doesn't look right. The outer check has `GlobalGrids` but the inner logic concerns the distance snap grid. Isn't that a `RelativeGrid`? What am I missing here?

peppy replied:
More like what did *I* miss.

peppy committed:
Fix incorrect catch grid specification

bdach committed:
Don't snap non-head slider nodes to distance grid during placement either

1cd69220ef640424b79f19784c557ac183333c2d only disabled snapping the
aforementioned nodes to distance grid for already-placed sliders.
`SliderPlacementBlueprint` has its own logic for placement, so the fix
needs to be mirrored there too.

bdach committed:
Merge branch 'master' into slider-control-point-no-distance-snap

bdach commented:
I've had to apply the same fix a second time to `SliderPlacementBlueprint`, as it's a disparate code path from `PathControlPointVisualiser` in https://github.com/ppy/osu/pull/23662/commits/e35201cb991aac8b27d68cb3dd49eb431d68d844. Looks good otherwise, though.
